### ISSUE TITLE
Possible memory leak in the parser.

### ISSUE THREAD

**Issue by blais** [beancount/beancount#124]
**[Original report](https://bitbucket.org/blais/beancount/issue/124) by Jakob Schnitzer (Bitbucket: [yagebu](https://bitbucket.org/yagebu), GitHub: [yagebu](https://github.com/yagebu)).**

----------------------------------------

In the process of investigating https://github.com/aumayr/fava/issues/247, I've found that the script below leads to a high memory usage. This leads me to believe the cause for the high memory usage of a long running `fava` process lies somewhere in the parser (if the file is read from cache, the memory usage is basically constant). Are there any global objects that might not be discarded by the garbage collection?

    import os

    os.environ['BEANCOUNT_DISABLE_LOAD_CACHE'] = '1'

    from beancount import loader    # noqa

    def load_file():
        loader.load_file(os.environ['BEANCOUNT_FILE'])

    for i in range(100):
        load_file()

**Comment by blais** [comment]
**Original comment by Jakob Schnitzer (Bitbucket: [yagebu](https://bitbucket.org/yagebu), GitHub: [yagebu](https://github.com/yagebu)).**

----------------------------------------

Sorry, I hit enter before my cursor was in the text input field, that's why I sent the empty issue at first. I've updated the description now.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

It's likely you're right. I'll have to investigate further.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

There's definitely something going on.
 

```
#!

mandarine [hg|default]:~/p/repos/beancount1$ BEANCOUNT_DISABLE_LOAD_CACHE=1   valgrind --tool=memcheck  --suppressions=$HOME/src/python/cpython/Misc/valgrind-python.supp ~/src/python/vginstall/bin/python3 /tmp/yagebu.py                                                                                                                                                  
==5696== Memcheck, a memory error detector
==5696== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==5696== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==5696== Command: /home/blais/src/python/vginstall/bin/python3 /tmp/yagebu.py
==5696== 
==5696== 
==5696== HEAP SUMMARY:
==5696==     in use at exit: 25,450,558 bytes in 227,960 blocks
==5696==   total heap usage: 2,383,017 allocs, 2,155,057 frees, 298,743,976 bytes allocated
==5696== 
==5696== LEAK SUMMARY:
==5696==    definitely lost: 0 bytes in 0 blocks
==5696==    indirectly lost: 0 bytes in 0 blocks
==5696==      possibly lost: 9,364,218 bytes in 60,768 blocks
==5696==    still reachable: 16,086,340 bytes in 167,192 blocks
==5696==         suppressed: 0 bytes in 0 blocks
==5696== Rerun with --leak-check=full to see details of leaked memory
==5696== 
==5696== For counts of detected and suppressed errors, rerun with: -v
==5696== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

```

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Fixed issue #124, a leak in the grammar parser which would leak the entire set of entries on every instance of parse.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Try this now, that's likely the big one. I did some hunting down using valgrind and this tool call "pympler" which turned out to be particularly easy to use and useful.
